stannp/stannp.go

```diff
@@ -141,12 +144,71 @@ func (s *Stannp) post(inputReader io.Reader, inputURL string) (*http.Response, *
 
 	res, err := s.client.Do(req)
 	if err != nil {
-		return nil, util.BuildError(500, fmt.Sprintf("error sending req [%+v]", req), false)
+		return nil, util.BuildError(500, fmt.Sprintf("error sending req [%+v]", req))
 	}
 
 	return res, nil
 }
 
+func (s *Stannp) BytesToPDF(data []byte) (*os.File, *util.APIError) {
+	tmpFile, err := ioutil.TempFile("", "example.*.pdf")
+	if err != nil {
+		return nil, util.BuildError(500, err.Error())
+	}
+
+	if _, writeErr := tmpFile.Write(data); writeErr != nil {
+		closeErr := tmpFile.Close()
+		if closeErr != nil {
+			return nil, util.BuildError(500, closeErr.Error())
+		}
+		return nil, util.BuildError(500, writeErr.Error())
+	}
+
+	return tmpFile, nil
+}
+
+func (s *Stannp) DownloadPDF(pdfURL string) (*letter.PDFRes, *util.APIError) {
+	if !strings.HasPrefix(pdfURL, PDFURLPrefix) {
+		return nil, util.BuildError(400, fmt.Sprintf("pdfURL must begin with [%s]. your input was [%s]", PDFURLPrefix, pdfURL))
+	}
+
+	fileURL, err := url.Parse(pdfURL)
+	if err != nil {
+		log.Fatal(err)
+	}
+	path := fileURL.Path
+	segments := strings.Split(path, "/")
+	fileName := segments[len(segments)-1]
+
+	client := http.Client{
+		CheckRedirect: func(r *http.Request, via []*http.Request) error {
+			r.URL.Opaque = r.URL.Path
+			return nil
+		},
+	}
+
+	resp, err := client.Get(pdfURL)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	byteArray, err := ioutil.ReadAll(resp.Body)
```

Comment:
Would be nice if we didn't have to read the entire contents of the PDF file into memory. If instead a `io.ReadCloser` was returned, the caller has more flexibility with handling reading, which might be piping into a writer.

Reply:
Nice, this pattern is pretty neat. Would we want to use an `io.NopCloser` for additional flexibility as per this [blog](https://medium.com/@xoen/golang-read-from-an-io-readwriter-without-loosing-its-content-2c6911805361)?

Reply:
@z-cran we want to properly propagate the `Close` on the `*http.Request.Body`. I would think you could just return the request body directly.
Since the caller is going to be doing the reading, you need to leave it to the caller to properly close when the reading is done.

Reply:
awesome feedback y'all - thanks! implemented.